### PR TITLE
Inference context annotations

### PR DESCRIPTION
**Pull request by nickdrozd**
## Description

Add type annotations pertaining to inference contexts.

1. The file `context.py` is mostly typed.
2. Most if not all inference context args are annotated.
3. Two minor code errors exposed by the type checker are fixed.

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |
| ✓   | :hammer: Refactoring   |

## Related Issue

This fixes a code error. I don't know if it was associated with any known bug.

### CODE REVIEW

File: astroid/context.py [hunk 1]
```diff
@@ -174,7 +178,7 @@ def copy_context(context: InferenceContext | None) -> InferenceContext:
     return InferenceContext()
 
 
-def bind_context_to_node(context, node):
+def bind_context_to_node(context: InferenceContext | None, node) -> InferenceContext:
```
Comment (DanielNoord) [inline astroid/context.py:181]: Could you remove the docstring typing here?

File: astroid/context.py [hunk 2]
```diff
@@ -104,7 +108,7 @@ def inferred(self) -> _InferenceCache:
         """
         return _INFERENCE_CACHE
 
-    def push(self, node):
```
Comment (DanielNoord) [inline astroid/context.py:107]: Can you remove `rtype` here as well?

File: astroid/context.py [hunk 3]
```diff
@@ -72,7 +76,7 @@ def __init__(self, path=None, nodes_inferred=None):
 
         e.g. the bound node of object.__new__(cls) is the object node
         """
-        self.extra_context = {}
+        self.extra_context: dict[NodeNG, InferenceContext] = {}
```
Comment (DanielNoord) [inline astroid/context.py:79]: Have you checked that this can't be `Instance` or its parent `Proxy`? In general, docstring typing is incorrect 😓

File: astroid/brain/brain_builtin_inference.py [hunk 4]
```diff
@@ -693,7 +693,7 @@ def infer_issubclass(callnode, context=None):
     return nodes.Const(issubclass_bool)
 
 
-def infer_isinstance(callnode, context=None):
+def infer_isinstance(callnode, context: InferenceContext | None = None):
     """Infer isinstance calls
 
     :param nodes.Call callnode: an isinstance call
```
Comment (Pierre-Sassoulas) [inline astroid/brain/brain_builtin_inference.py:699]: The docstring typing is duplicated and can be removed imo. (There are other instances of this)

File: astroid/context.py [hunk 5]
```diff
@@ -42,7 +42,11 @@ class InferenceContext:
 
     max_inferred = 100
 
-    def __init__(self, path=None, nodes_inferred=None):
+    def __init__(
+        self,
+        path: set[tuple[NodeNG, str | None]] | None = None,
+        nodes_inferred: list[int] | None = None,
```
Comment (DanielNoord) [inline astroid/context.py:48]: Why is this a list 😅 Seems so illogical, oh well..

File: astroid/context.py [hunk 6]
```diff
@@ -42,7 +42,11 @@ class InferenceContext:
...
+    ):
```
Comment (DanielNoord) [inline astroid/context.py:49]:
```suggestion
    ) -> None:
```

File: astroid/context.py [hunk 7]
```diff
@@ -42,7 +42,11 @@ class InferenceContext:
 
     max_inferred = 100
 
-    def __init__(self, path=None, nodes_inferred=None):
+    def __init__(
+        self,
+        path: set[tuple[NodeNG, str | None]] | None = None,
```
Comment (DanielNoord) [inline astroid/context.py:47]: Where do you base this on?

**Review by DanielNoord — CHANGES_REQUESTED** [review]
Only spotted this somewhere along the review: please review any docstring typing that is now explicit. It makes sure we don't have to keep two separate things up to date!

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
Good job finding those two bugs !

File: astroid/context.py (hunk 3 above)
Comment (nickdrozd) [reply astroid/context.py:79]: Yes, I will cut this one.

File: astroid/context.py (hunk 6 above)
Comment (nickdrozd) [reply astroid/context.py:49]: Isn't `__init__` exempt from return typing?

File: astroid/context.py (hunk 7 above)
Comment (nickdrozd) [reply astroid/context.py:47]:
```
        :type: set(tuple(NodeNG, optional(str)))

        Path of visited nodes and their lookupname

        Currently this key is ``(node, context.lookupname)``
```

File: astroid/context.py (hunk 3 above)
Comment (nickdrozd) [reply astroid/context.py]: Changed it to `Any` instead.

File: astroid/context.py (hunk 7 above)
Comment (DanielNoord) [reply astroid/context.py:47]: It's very likely that this is indeed `SuccesfulInferenceResult` instead of `NodeNG`. Let's revert this and keep it without the typing for now. The PR is really useful anyway!

File: astroid/context.py (hunk 6 above)
Comment (DanielNoord) [reply astroid/context.py:49]: Ah of course. Just force of habit for me, feel free to ignore and resolve!

File: astroid/context.py [hunk 8]
```diff
@@ -64,24 +67,20 @@ def __init__(self, path=None, nodes_inferred=None):
         """
         self.callcontext: CallContext | None = None
         """The call arguments and keywords for the given context."""
-        self.boundnode = None
+        self.boundnode: NodeNG | None = None
```
Comment (DanielNoord) [inline astroid/context.py:70]: Again, are we sure that this is not more than `NodeNG`? I'd prefer to do such changes in a separate PR.

File: astroid/context.py [hunk 9]
```diff
@@ -64,24 +67,20 @@ def __init__(self, path=None, nodes_inferred=None):
...
         """
-        :type: optional[NodeNG]
-
         The bound node of the given context
 
         e.g. the bound node of object.__new__(cls) is the object node
         """
-        self.extra_context = {}
+        self.extra_context: dict[Any, InferenceContext] = {}
```
Comment (DanielNoord) [inline astroid/context.py:76]: Let's not use `Any`. Since `mypy` isn't running anyway the typing also doesn't give us much so it might be better to just keep the docstring for now until we can be 100% sure we get this correctly.

**Review by DanielNoord — CHANGES_REQUESTED** [review]

File: astroid/context.py (hunk 9 above)
Comment (nickdrozd) [reply astroid/context.py:80]: I'd like to keep this one because for some reason Mypy specifically calls it out:

```
astroid/context.py:78: error: Need type annotation for "extra_context" (hint: "extra_context: Dict[<type>, <type>] = ...")  [var-annotated]
```

So when trying to do more typing, it adds noise.

File: astroid/context.py (hunk 7 above)
Comment (nickdrozd) [reply astroid/context.py]: I changed `NodeNG` to `Any`. Even with that hole, this a pretty specific type to verify.

I have a feeling that this pattern is going to come up all over the place, where a docstring says it's `NodeNG`, but in reality it's `NodeNG | Proxy | Base | Etc`. Being able to verify the compound types without having to untangle all the node-like types would be easier.

Comment (DanielNoord) [reply astroid/context.py]: I don't really understand your last comment.

I agree it would be good to untangle where the docstring annotation is incorrect but I prefer to do so in separate PRs as checking that is so involved. If we save that change for a small follow-up PR we can focus on getting that right while we can now already merge the `context: InferenceContext | None` stuff.

**Comment by nickdrozd** [comment]
@DanielNoord I took out the partial annotations. I'll put them into a separate PR.

**Review by DanielNoord — APPROVED** [review]

**Comment by DanielNoord** [comment]
THanks a lot @nickdrozd :heart:

**Review by Pierre-Sassoulas — APPROVED** [review]
Nice 🎉 !
